fix: SDK crashes due to missing error code property in `ParseNetworkResponse.data`

mbfakourii opened the pull request:
### New Pull Request Checklist
- [x] I am not disclosing a [vulnerability](https://github.com/parse-community/Parse-SDK-Flutter/security/policy).
- [x] I am creating this PR in reference to an [issue](https://github.com/parse-community/Parse-SDK-Flutter/issues?q=is%3Aissue).

### Issue Description
sometimes in buildErrorResponse ```code``` not exist in ```apiResponse.data``` and ```json.decode``` can't parse it!
but statusCode exist in ParseNetworkResponse and can be taken from this variable.

Related issue: https://github.com/parse-community/Parse-SDK-Flutter/issues/799
Closes: https://github.com/parse-community/Parse-SDK-Flutter/issues/799

### Approach
In some times when the ```code``` is not available in ```ParseNetworkResponse.data```, the ```statusCode``` is used in the ```ParseNetworkResponse.statusCode``` variable.

### TODOs before merging
- [ ] Add tests
- [ ] Add changes to documentation (guides, repository pages, in-code descriptions)
- [ ] A changelog entry

mtrezza commented:
Which side effects could that change have?

mbfakourii commented:
> What side effects could that change have? The way of building the error response with `responseData[keyCode]` has worked so far it seems?

I did various tests and found out that sometimes the ```code``` is not sent. That's why I put the null checker ```??``` for when it is not available, and when the ```code``` is not sent, it uses ```ParseNetworkResponse.statusCode```.

mtrezza commented:
So the underlying issue is that Parse Server sometimes doesn't send that code?

mbfakourii commented:
> So the underlying issue is that Parse Server sometimes doesn't send that code?

yes for example 
if ```appID``` is wrong and try for get ```health```
```https://parseapi.back4app.com/health```
Json received without ```code``` is like this
```json
{"error":"unauthorized"}
```

mtrezza commented:
Got it, I'm a bit careful about introducing this ambiguity. It basically means that a developer cannot be sure whether the code is a parse error code or a http response code, right?

For example, if the error code is 400 and Parse Server also has a Parse Error code 400, how should a developer interpret this, and how should error handling logic in an app deal with this?

mbfakourii commented:
> Got it, I'm a bit careful about introducing this ambiguity. It basically means that a developer cannot be sure whether the code is a parse error code or a http response code, right?
> 
> For example, if the error code is 400 and Parse Server also has a Parse Error code 400, how should a developer interpret this, and how should error handling logic in an app deal with this?
> 
> Maybe an alternative would be to use a distinct generic Parse Error code if no specific error code is present.
> 
> For example the [docs](https://docs.parseplatform.org/js/guide/#other-issues) already offer `OtherCause` (-1): "An unknown error or an error unrelated to Parse occurred."

It is true that the developer is not able to understand the meaning of the code.
I use ```-1``` instead of ```ParseNetworkResponse.statusCode```

mtrezza commented on packages/dart/lib/src/objects/response/parse_error_response.dart:
```diff
@@ -5,7 +5,8 @@ ParseResponse buildErrorResponse(
-  response.statusCode = responseData[keyCode];
+      code: responseData[keyCode] ?? -1,
+      message: responseData[keyError].toString());
+  response.statusCode = responseData[keyCode] ?? -1;
```
Do we have a list of Parse Errors in this SDK anywhere? These error codes are usually defined somewhere. See for example [ParseError](https://github.com/parse-community/Parse-SDK-JS/blob/alpha/src/ParseError.js) in the Parse JS SDK.

If the Flutter SDK doesn't have that, could you please centrally defined at least the `-1` error code with the same name as in the Parse JS SDK, and use that reference instead of directly returning `-1`.

mbfakourii replied:
In the Flutter SDK, we have a list of errors in this [file](https://github.com/parse-community/Parse-SDK-Flutter/blob/d5f0657dd06fbb4ad6c7890303e946d72d39e773/packages/dart/lib/src/objects/parse_error.dart#L17), but the variable is private and cannot be used in other classes. I will try to improve this section.

mtrezza replied:
I think it should be public anyway, because a developer may want to reference the error code somewhere in their code to identify the error.

mbfakourii replied:
> Amazing! Could you please preserve the numeric codes as well? It seems that in the commit they were removed.

all numeric codes are linked to const keys. like [this](https://github.com/parse-community/Parse-SDK-Flutter/blob/60a82288aa329b0537b5d936e6c5785837294be7/packages/dart/lib/src/objects/parse_error.dart#L228).
what codes do you mean ?

mtrezza replied:
Looking at it now it looks fine, I think I only saw a truncated version when looking at the changes previously.

mbfakourii commented:
Changed and added const error code based on [Parse-SDK-JS](https://github.com/parse-community/Parse-SDK-JS/blob/5c140299109d2850982155ac1e1b907e1895fde0/src/ParseError.js)

mtrezza commented on packages/dart/lib/src/objects/parse_error.dart:
```diff
@@ -13,58 +13,279 @@ class ParseError {
-    -1: 'UnknownError',
-
-    // SDK errors / Errors
-    1: 'No Results',
```
It seems that these error codes (which you already removed) were actually some Flutter SDK-custom codes (1, 2) and a HTTP response code (400).

Have these 3 codes been used anywhere? If so, this is a breaking change, at least for code 1 which was described as "No Results" but is now "internalServerError".

mtrezza replied:
I don't think it makes sense to return them, the 400 code is incorrect anyway, because it mixes a HTTP response code into the Parse Error codes. If you are 100% sure they weren't used anywhere we can just remove them. Did you also do a simple text search to be sure?

mbfakourii replied:
yes 
In addition, I saw error 400 in several issues, none of which were related to code 400 https://github.com/parse-community/Parse-SDK-Flutter/issues/487#issue-721516490  ,https://github.com/parse-community/Parse-SDK-Flutter/issues/487#issue-721516490 and https://github.com/parse-community/Parse-SDK-Flutter/issues/481#issue-719863736

contributors parse_error.dart are codes 1, 2 and 400 important?
@nstrelow 
@fischerscode

mbfakourii replied:
@mtrezza 
I suggest not to add codes unless there is a problem because I don't see any use of these codes in Parse Flutter!

mtrezza replied:
Sounds good

mtrezza reviewed (dismissed):
Looks good to me; unless we get contrary feedback regrading the 3 removed codes.

- Is the PR title correct?
- Could you add the changelog entry by copy/pasting the PR title + bump the version for the release?

mbfakourii commented:
> @mbfakourii You have been quite active recently and make some great contributions. Can I add you to the review team of the Parse Flutter SDK? You would get notified if there are PRs to review.

Yes, I will be happy to add

 and title added in changelog

mtrezza commented on packages/flutter/pubspec.yaml:
```diff
@@ -1,6 +1,6 @@
 name: parse_server_sdk_flutter
 description: Flutter plugin for Parse Server, (https://parseplatform.org), (https://back4app.com)
-version: 3.1.3
+version: 3.1.4
```
Why is the flutter version being bumped if the change is only in the dart package?

See https://github.com/parse-community/Parse-SDK-Flutter/commit/7fb7ce67d5bea30d583f4b182303ee6a31214185 where we also only had a change in the dart package and only the dart version was bumped.

mbfakourii replied:
It was my mistake, I thought it was for dart

mbfakourii replied:
Returned to the previous version

mtrezza commented on packages/dart/lib/src/base/parse_constants.dart:
```diff
@@ -1,7 +1,7 @@
 
 // Library
-const String keySdkVersion = '3.1.2';
+const String keySdkVersion = '3.1.4';
```
Did you manually change this version or is this auto-generated?

mbfakourii replied:
> Did you manually change this version or is this auto-generated?

manually

mtrezza replied:
We haven't changed that in the last release, maybe by mistake?

I think we should create a list where the version has to be changed for a new release.

mbfakourii replied:
> We haven't changed that in the last release, maybe by mistake?

Yes, in the previous version, I didn't know exactly that I had to change the parse_constants file

> I think we should create a list where the version has to be changed for a new release. So for dart it's in:
> * packages/dart/lib/src/base/parse_constants.dart
> * packages/dart/pubspec.yaml
> 
> ?

Yes, it's a good idea

mtrezza replied:
Why is this variable here needed at all?

mbfakourii replied:
This variable is sent to the server as the version of the requester in a header in keys ```parseVersion``` and ```user-agent```

and It is not a critical variable, but it is good to change it

mtrezza replied:
So this was actually bug in the previous version then. Is there any way we could get the package version from the package itself instead of manually maintaining this var?

mbfakourii replied:
> So this was actually bug in the previous version then. 

yes
> Is there any way we could get the package version from the package itself instead of manually maintaining this var?

Yes, but you have to use other packages [package_info_plus](https://pub.dev/packages/package_info_plus) and this package does not support the Dart version, it only supports Flutter!

mbfakourii replied:
@mtrezza 
Is it possible to Merge this PR? I need these codes in another PRs.

mtrezza replied:
Sure, let' keep this in mind, maybe we can find a way to replace that extra var in the future...

mtrezza commented on packages/dart/CHANGELOG.md:
```diff
@@ -1,3 +1,9 @@
+
+### Bug Fixes
+
+* fix: SDK crashes due to missing error code property in `ParseNetworkResponse.data` ([#799](https://github.com/parse-community/Parse-SDK-Flutter/issues/799))
```
```suggestion
* SDK crashes due to missing error code property in `ParseNetworkResponse.data` ([#799](https://github.com/parse-community/Parse-SDK-Flutter/issues/799))
```

mtrezza commented on packages/flutter/pubspec.yaml:
```diff
@@ -10,7 +10,7 @@ dependencies:
     sdk: flutter
 
-  parse_server_sdk: ^3.1.2
+  parse_server_sdk: ^3.1.3
```
I think this also needs to be reverted, otherwise we'd have to release a new flutter version?

mbfakourii replied:
No, there is no need Returned to the previous version

mtrezza commented:
Off-topic; if you're interested please feel free to join our Slack workspace. There are actually a few Flutter related developer questions that came in today, and you seem to be a Flutter expert, so this could be a good match.

mtrezza commented on packages/dart/CHANGELOG.md:
```diff
@@ -1,3 +1,9 @@
+## [3.1.4](https://github.com/parse-community/Parse-SDK-Flutter/compare/dart-3.1.3...dart-3.1.4) (2022-12-10)
```
```suggestion
## [3.1.4](https://github.com/parse-community/Parse-SDK-Flutter/compare/dart-3.1.3...dart-3.1.4) (2022-12-14)
```

mtrezza reviewed (approved):
Looks good!
